회장 위임 기능 반영

Lee-sh98 opened the pull request:
## What is this Pull Request about? 💬
- 회장 위임 기능의 구현


## Key Changes 🔑
- 회장 위임 기능에 필요한 controller, service 구현
- AdminPage 테스트 생성


## To Reviewers 🙏
- @binary-ho
- @ghdeo

Lee-sh98 committed:
feat: add controller for president-delegate feature (#10)

Lee-sh98 committed:
feat: add president-delegate feature (#10)

Lee-sh98 committed:
test: add AdminPageTest (#10)

binary-ho commented on src/main/java/hicc/toy_project/controller/AdminPageApiController.java:
```diff
@@ -74,5 +76,18 @@ public boolean approve(@RequestBody AdminPageRequest request) {
+    public ResponseEntity<Void> delegate(@RequestBody AdminPageRequest request){
+        adminPageService.delegate(request);
 
+        return new ResponseEntity<>(HttpStatus.OK);
```
ResponseEntity에 Status Code를 기입하지 않으면 어떤 응답을 받게 될까요?

Lee-sh98 replied:
#### Status Code를 기입하지 않을 수 없습니다.
ResponseEntity는 HttpEntity를 확장하고 있습니다.

HttpEntity
- headers (HttpHeaders)
- body (T)

ResponseEntity<T>는 위 HttpEntity에 Object타입의 status를 갖고 있습니다.
이 ResponseEntity의 생성자는 5가지로 다음과 같습니다.
- ResponseEntity(HttpStatusCode status)
- ResponseEntity(@Nullable T body, HttpStatusCode status)
- ResponseEntity(MultiValuMap<String, String> headers, HttpStatusCode status)
- ResponseEntity(@Nullable T body, @Nullable MultiValueMap<String, String> headers, HttpStatusCode status)
- ResponseEntity(@Nullable T body, @Nullable MultiValueMap<String, String> headers, int rawStatus)

생성자들은 HttpStatusCode를 받거나 rawStatus를 받고, 이후this()를 통해 private 생성자를 호출하여 값을 설정합니다.
따라서 status code를 기입하지않을 수 없습니다.

ghdeo commented on src/main/java/hicc/toy_project/service/AdminPageService.java:
```diff
@@ -156,4 +156,24 @@ public boolean approve(AdminPageRequest request) {
+                        new CustomException(ErrorCode.MEMBER_NOT_FOUND)
+                );
+
+        if (!isPresident(request.getId())) {
```
#9 에서 반영하신 changeRole() 로직 개선과 비슷한 맥락인데 isPresident를 통해 request가 합리적인지 판단하는 로직을 코드 최상단에 올리리면 request가 valid하지 않을 때에도 repository에서 굳이 currentPresident와 nextPresident를 꺼내올 필요가 없을 것 같아요.

Lee-sh98 replied:
수정하였습니다.

ghdeo commented on src/test/java/hicc/toy_project/adminpage/AdminPageTest.java:
```diff
@@ -0,0 +1,77 @@
+        Member nextPresident = getMember(testTargetId);
+
+        currentPresident.setRole(Role.GENERAL);
+        nextPresident.setRole(Role.PRESIDENT);
```
Member의 setter에 대한 테스트로 보이네요.
AdminPageService의 delegate메서드에 대한 테스트거나 적어도 Member의 updateRole메서드에 대한 테스트여야 할 것 같아요.

Lee-sh98 replied:
수정하였습니다.

Lee-sh98 committed:
feat: 회장 위임 기능 테스트 수정 (#10)

Lee-sh98 committed:
feat: Member 생성자 수정, 멤버 함수 수정 (#10)

setter 삭제로 인한 생성자 및 멤버 함수 수정

Lee-sh98 committed:
feat: MemberAbstract 수정 (#10)

Setter 삭제
멤버 변수 protected 적용

Lee-sh98 committed:
feat: AdminPageRequest 빌더 생성 (#10)

Lee-sh98 committed:
refactor: 회장 검증 로직 순서 변경 (#10)

binary-ho commented on src/main/java/hicc/toy_project/controller/dto/AdminPageRequest.java:
```diff
@@ -1,23 +1,23 @@
 
 @Getter
-@NoArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
AccessLevel을 변경한 이유가 궁금합니다

Lee-sh98 replied:
request dto의 무분별한 생성을 막기 위해 접근제한을 강화하면서 추후 클래스를 extend하여 사용할 때를 대비해 설정하였습니다.

binary-ho commented on src/test/java/hicc/toy_project/adminpage/AdminPageTest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    @Test
+    @DisplayName("회장_위임_기능")
```
테스트에 대한 이름을 지을 때, 좀 더 구체적이였으면 좋겠습니다.
~~해서 ~~할 수 있다. 와 같이 상세한 설명이 있는 편이 다른 사람이 테스트를 볼 때도 어떤 기능을 위한 테스트인지 좀 더 직관적으로 파악할 수 있을 것 같습니다

binary-ho replied:
예를 들면 여기선 회장 권한을 위임하고, 기존 회장을 일반 회원으로 만든다 뭐 이런 식이면 다른 사람이 코드를 볼 때 좀 더 명확할 것 같습니다

Lee-sh98 replied:
테스트 이름을 명확히 수정하겠습니다.

Lee-sh98 committed:
feat: AdminPageRequest 빌더 생성자에 등급 추가 (#10)

Lee-sh98 committed:
feat: AdminPageRequest 빌더 생성자에 승인여부 추가 (#10)

Lee-sh98 committed:
test: AdminPage 테스트 추가 (#10)

- 회장 위임 기능
  - 정상 동작
- 회원 등급 수정 기능
  - 정상 동작
  - 회장 본인 등급 수정 예외
  - 회장으로의 등급 수정 예외
  - 게스트로 등급 수정 예외
  - 같은 등급으로 수정 예외
- 회원 제명 기능
  - 정상 동작
  - 회장 본인 제명 예외
- 회원 가입 승인 기능
  - 정상 승인
  - 정상 반려
  - 지원자가 아닌 회원 승인 요청 예외

Lee-sh98 committed:
Merge branch 'develop' into feature/president-resign

Lee-sh98 committed:
test: static import assertj(#10)

Lee-sh98 committed:
Merge branch 'feature/president-resign' of https://github.com/Lee-sh98/toy_project into feature/president-resign
